File: my-folder/0054-spiral-matrix/solution.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int top=0, bottom=matrix.size()-1;
        int left=0, right=matrix[0].size()-1;
        int dir=0;
        vector<int>res;

        while(top<=bottom and left<=right)
        {
            if(dir==0)
            {
                for(int i=left; i<=right; i++)
                res.push_back(matrix[top][i]);
                top++;
            }
            else if(dir==1)
            {
                for(int j=top; j<=bottom; j++)
                res.push_back(matrix[j][right]);
                right--;
            }
            else if(dir==2)
            {
                for(int k=right; k>=left; k--)
                res.push_back(matrix[bottom][k]);
                bottom--;
            }
            else if(dir==3)
            {
                for(int l=bottom; l>=top; l--)
                res.push_back(matrix[l][left]);
                left++;
            }
            dir=(dir+1)%4;
        }
        return res;

    }
};

File: my-folder/0054-spiral-matrix/solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        top, bottom = 0, len(matrix)-1
        left, right = 0, len(matrix[0])-1
        dir = 0
        res = []

        while top<=bottom and left<=right:
            if dir==0:
                for i in range(left, right+1):
                    res.append(matrix[top][i])
                top=top+1
            elif dir==1:
                for j in range(top, bottom+1):
                    res.append(matrix[j][right])
                right=right-1
            elif dir==2:
                for k in range(right, left-1, -1):
                    res.append(matrix[bottom][k])
                bottom=bottom-1
            elif dir==3:
                for l in range(bottom, top-1, -1):
                    res.append(matrix[l][left])
                left=left+1
            dir=(dir+1)%4
        return res

